UTM_GLOBAL_POSITION message

eyeam3 opened the pull request:
Introduction of UTM_GLOBAL_POSITION as discussed in https://github.com/mavlink/rfcs/pull/10.

auturgy commented on message_definitions/v1.0/common.xml:
```diff
@@ -2998,6 +2998,45 @@
+      <entry value="3" name="UTM_FLIGHT_STATE_AIRBORNE">
+        <description>UAS airborne.</description>
+      </entry>
+    </enum>
```
Suggest additional state "UTM_FLIGHT_STATE_EMERGENCY".

eyeam3 replied:
Good idea, I will add the state.

auturgy commented on message_definitions/v1.0/common.xml:
```diff
@@ -2998,6 +2998,45 @@
+      <entry value="128" name="UTM_DATA_AVAIL_FLAGS_NEXT_WAYPOINT_AVAILABLE">
+        <description>The fields next_lat, next_lon and next_alt contain valid data.</description>
+      </entry>
+    </enum>
```
How the flags have been split up doesn't flow logically to me.  In fact some groupings make little sense at all.
Needs rework.

eyeam3 replied:
Can you please elaborate on your comment a bit more? What split would you propose instead?

auturgy replied:
When I first read through this, my brain assumed v_acc and h_acc were accelerations.  Which made no sense.  Accuracy makes more sense :)
There is relevant discussion re accuracy/uncertainty here: https://github.com/mavlink/mavlink/issues/624

eyeam3 replied:
Maybe the naming is unfortunate, I will check other messages and rename to stay consistent in MAVLink.
The discussion is very interesting, in particular the categories. I prefer to publish the information as standard deviation since it is more precise and could be converted to the categories afterwards.

auturgy commented on message_definitions/v1.0/common.xml:
```diff
@@ -4821,6 +4860,29 @@
+      <!-- This message is work-in-progress it can therefore change, and should NOT be used in stable production environments -->
+      <description>The global position resulting from GPS and sensor fusion.</description>
+      <field type="uint64_t" name="time" units="us">Time of applicability of position (microseconds since UNIX epoch).</field>
+      <field type="uint8_t[18]" name="uas_id">Unique UAS ID.</field>
```
Existing ADS-B identifier uses a single uint32_t for the ICAO address.  Acknowledging the likely need for a very large number of identifiers, what is the specific use case driving this to be an array?  Some justification is required, as this is a critical field that won't be easy to change once the use of this message proliferates.

eyeam3 replied:
The choice of an array with 18 byte is motivated by the utilization of UUID from the chip (which is up to 128 bit on some architectures) plus 2 byte to distinguish different architectures. The reasoning is discussed in this issue: https://github.com/PX4/Firmware/issues/7891#issuecomment-326608612.

auturgy replied:
OK.  The obvious logical flaw is that the UUID is unique to the MCU of the flight controller, not the air vehicle.  Whilst it sort of works conceptually for RTF manufacturers, this not a good idea.
It changes ID's unnecessarily if you change the flight controller on a vehicle (not uncommon, although infrequent), it shifts ID's if you move the flight controller to another vehicle (again, infrequent but not uncommon), and it isn't domain controllable.
Obviously as a 128bit x 2byte number we aren't going to run out of ID's, but using the UUID is a convenient hack that forgets the actual purpose of the field: a unique identifier for the UAV.  Needs more discussion, and input from the relevant regulators.

eyeam3 replied:
You are absolutely correct about the flaws concerning the UUID and it might not be used in the final UTM system. As far as I know, in the foreseeable future there is no global regulation that organizes unique identifiers for UAVs, which in turn means we cannot directly adapt such a standard. Using the UUID would give us a unique identifier without requiring the developer or pilot to change anything, move the discussion forward and enable various demonstrations. 
It is important to note that the definition is considered as "work in progress" also after the merge and the message should be used with care, as there will be modifications to acknowledge outside requirements. 
To come back to the size of the field, an array with 18 byte should be large enough to accommodate any type of identifier and should have a more consistent interface compared to a uint32_t. Yet, the size of the array might change when the topic evolves.
I understand that such modifications can break existing implementations and we have to avoid to scatter the identifier over to many different types.

eyeam3 commented on message_definitions/v1.0/common.xml:
```diff
@@ -4821,6 +4860,29 @@
+      <field type="int16_t" name="vz" units="cm/s">Ground Z Speed (Altitude, positive down), expressed as m/s * 100.</field>
+      <field type="uint32_t" name="h_acc" units="mm">Horizontal Position uncertainty in meters * 1000. (standard deviation)</field>
+      <field type="uint32_t" name="v_acc" units="mm">Altitude uncertainty in meters * 1000. (standard deviation)</field>
+      <field type="uint32_t" name="vel_acc" units="mm/s">Speed uncertainty in m/s * 1000. (standard deviation)</field>
```
I can make them uint16.

auturgy replied:
Just a further comment: having the measurement and it's related uncertainty in different units could result in implementation errors.  Not absolutely necessary, but consistency is a good thing.

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -4821,6 +4860,29 @@
       <field type="float[5]" name="delta" units="s">Bezier time horizon, set to NaN if velocity/acceleration should not be incorporated</field>
       <field type="float[5]" name="pos_yaw" units="rad">Yaw, set to NaN for unchanged</field>
     </message>
+    <message id="334" name="UTM_GLOBAL_POSITION">
```
Does it make sense to leave a few empty ids in case we need some new trajectory messages and want them adjacent? We don't  have any rules on this, but it is obviously easier for readers if associated messages are together, and the available message space in MAVLink 2 is large.

eyeam3 replied:
I am open to change the id and would follow the best practice from MAVLink. Do we have a recommendation here?

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -4821,6 +4860,29 @@
+      <field type="int32_t" name="relative_alt" units="mm">Altitude above ground in meters * 1000.</field>
+      <field type="int16_t" name="vx" units="cm/s">Ground X Speed (Latitude, positive north), expressed as m/s * 100.</field>
+      <field type="int16_t" name="vy" units="cm/s">Ground Y Speed (Longitude, positive east), expressed as m/s * 100.</field>
+      <field type="int16_t" name="vz" units="cm/s">Ground Z Speed (Altitude, positive down), expressed as m/s * 100.</field>
```
not a big deal, but think about capitalisation here - it is odd and inconsistent.
ie

> Ground X Speed (Latitude, positive north), expressed as m/s * 100. 

could be

> Ground X speed (latitude, positive north), expressed as m/s * 100.

Similarly "Horizontal Position uncertainty in meters * 1000." should proabably be "Horizontal **p**osition uncertainty in meters * 1000."

eyeam3 replied:
Noted, I will adjust the capitalization.

thomasgubler commented:
@auturgy could you join the next mavlink devcall? https://mavlink.io/en/about/support.html#dev_call

auturgy commented:
@thomasgubler I’ve been to almost every one :)
I can’t guarantee the next call - I’ll be traveling - but will try.
I attend the first call. Feel free to tag me in the minutes if I don’t make it.

WickedShell commented on message_definitions/v1.0/common.xml:
```diff
@@ -4821,6 +4860,29 @@
+      <field type="uint32_t" name="h_acc" units="mm">Horizontal Position uncertainty in meters * 1000. (standard deviation)</field>
+      <field type="uint32_t" name="v_acc" units="mm">Altitude uncertainty in meters * 1000. (standard deviation)</field>
+      <field type="uint32_t" name="vel_acc" units="mm/s">Speed uncertainty in m/s * 1000. (standard deviation)</field>
+      <field type="int32_t" name="next_lat" units="degE7">Next waypoint, Latitude (WGS84), in degrees * 1E7.</field>
```
No DNU value if there isn't a target waypoint?

WickedShell replied:
Missed the flags.

WickedShell commented on message_definitions/v1.0/common.xml:
```diff
@@ -4821,6 +4860,29 @@
+      <field type="int32_t" name="next_alt" units="mm">Next waypoint, Altitude (WGS84), in meters * 1000.</field>
+      <field type="uint16_t" name="update_rate">Seconds * 1E2 until next update. Set to 0 if unknown or in data driven mode.</field>
+      <field type="uint8_t" name="flight_state">The flight state defined in enum \verb|UTM_FLIGHT_STATE|.</field>
+      <field type="uint8_t" name="flags">Bitwise OR combination of the enums \verb|UTM_DATA_AVAIL_FLAGS|.</field>
```
Tag the field with `enum="UTM_DATA_AVAIL_FLAGS"`

eyeam3 commented:
@hamishwillee I addressed the comments from this PR and from the MAVLink call and rebased onto current master.

Please have a look and let me know if there is anything I missed.

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -4821,6 +4866,29 @@
+      <field type="uint8_t[18]" name="uas_id">Unique UAS ID.</field>
+      <field type="int32_t" name="lat" units="degE7">Latitude (WGS84), in degrees * 1E7.</field>
+      <field type="int32_t" name="lon" units="degE7">Longitude (WGS84), in degrees * 1E7.</field>
+      <field type="int32_t" name="alt" units="mm">Altitude (WGS84), in meters * 1000.</field>
```
For `units` attributes generally we shouldn't need to state what units are in the description. So here 
```
<field type="int32_t" name="lat" units="degE7">Latitude (WGS84).</field>
      <field type="int32_t" name="lon" units="degE7">Longitude (WGS84).</field>
      <field type="int32_t" name="alt" units="mm">Altitude (WGS84).</field>
      <field type="int32_t" name="relative_alt" units="mm">Altitude above ground</field>
      <field type="int16_t" name="vx" units="cm/s">Ground X speed (latitude, positive north).</field>
...
```
@auturgy @amilcarlucas Are there any exceptions? 
The only one I think we might consider would be when the description states a standard that is always expected to be in a particular format. So if (for example) WGS84 mandates metres be used for altitude, then it might make sense to hightlight the non standard format. Don't think that applies here.

amilcarlucas replied:
I like the current descriptions, they do not include units, just directions and reference frames. I am fine with that.

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -4821,6 +4866,29 @@
+      <field type="int32_t" name="next_lon" units="degE7">Next waypoint, longitude (WGS84), in degrees * 1E7.</field>
+      <field type="int32_t" name="next_alt" units="mm">Next waypoint, altitude (WGS84), in meters * 1000.</field>
+      <field type="uint16_t" name="update_rate">Seconds * 1E2 until next update. Set to 0 if unknown or in data driven mode.</field>
+      <field type="uint8_t" name="flight_state" enum="UTM_FLIGHT_STATE">The flight state defined in enum \verb|UTM_FLIGHT_STATE|.</field>
```
@eyeam3 

> The flight state defined in enum \verb|UTM_FLIGHT_STATE|.

TO

> Flight state.

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -4821,6 +4866,29 @@
+      <field type="int32_t" name="next_alt" units="mm">Next waypoint, altitude (WGS84), in meters * 1000.</field>
+      <field type="uint16_t" name="update_rate">Seconds * 1E2 until next update. Set to 0 if unknown or in data driven mode.</field>
+      <field type="uint8_t" name="flight_state" enum="UTM_FLIGHT_STATE">The flight state defined in enum \verb|UTM_FLIGHT_STATE|.</field>
+      <field type="uint8_t" name="flags" enum="UTM_DATA_AVAIL_FLAGS">Bitwise OR combination of the enums \verb|UTM_DATA_AVAIL_FLAGS|.</field>
```
Hi @eyeam3 

description to 

> Bitmap of display availability flags.

Add attribute  `display="bitmask"`  (see [here for explanation](https://mavlink.io/en/guide/xml_schema.html#messages)

Do any fields have a default value you'd like to seed? You can specify that with `default=""`

eyeam3 replied:
I am not sure that this applies here. The purpose of this field is more for the logic of UTM and less for displaying purposes. Moreover, the available flags can change during the flight, e.g. when switching into mission mode, the UTM_DATA_AVAIL_FLAGS_NEXT_WAYPOINT_AVAILABLE flag would toggle.
And with regards to the explanation, it is not a checkbox that the user could select in the ground station, but defined by the FC software based on the flight state, available sensors and estimation.

Because of this, there are also no default values.

hamishwillee replied:
Fair enough on both of these, but as per other post, would be good if the description did not include the \verb ENum syntax.

eyeam3 replied:
Okay, makes sense.

amilcarlucas replied:
display="bitmask" also applies when data gets displayed, that way the GCS can display that multiple bits are active at the same time, and can explain which ones are.

hamishwillee commented:
@eyeam3 Thanks. A few more comments sorry!

eyeam3 commented:
@hamishwillee Thanks for the feedback. I addressed most of your additional comments, except for the `display` and `default` (see explanation above).

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -4821,6 +4866,29 @@
+      <field type="int32_t" name="next_alt" units="mm">Next waypoint, altitude (WGS84)</field>
+      <field type="uint16_t" name="update_rate">Seconds * 1E2 until next update. Set to 0 if unknown or in data driven mode.</field>
+      <field type="uint8_t" name="flight_state" enum="UTM_FLIGHT_STATE">Flight state</field>
+      <field type="uint8_t" name="flags" enum="UTM_DATA_AVAIL_FLAGS">Bitwise OR combination of the enum \verb|UTM_DATA_AVAIL_FLAGS|.</field>
```
@eyeam3 YOu missed this one :-). Specifically, enum is already referenced, and I don't think the system/rendering will like `\verb|UTM_DATA_AVAIL_FLAGS|.`

> Bitwise OR combination of the data available flags.

eyeam3 replied:
I changed the comment.

amilcarlucas commented on message_definitions/v1.0/common.xml:
```diff
@@ -4821,6 +4866,29 @@
+      <field type="int32_t" name="next_lat" units="degE7">Next waypoint, latitude (WGS84)</field>
+      <field type="int32_t" name="next_lon" units="degE7">Next waypoint, longitude (WGS84)</field>
+      <field type="int32_t" name="next_alt" units="mm">Next waypoint, altitude (WGS84)</field>
+      <field type="uint16_t" name="update_rate">Seconds * 1E2 until next update. Set to 0 if unknown or in data driven mode.</field>
```
is the update rate in centi-seconds ?

eyeam3 replied:
The field has only 16 bit and I don't want to increase its size. Given the size limit, I would like to leave the possibility to set a next update 10 minutes in the future and that would leave less precision for the next update, but I think this would be less important.

hamishwillee replied:
@amilcarlucas Were you asking in order to get correct units? Unfortunately we don't define this unit - perhaps we should? https://github.com/ArduPilot/pymavlink/blob/master/generator/mavschema.xsd#L69

In any case I think this reason is good.

amilcarlucas replied:
My comment is that the cs units needs to be added to pymavlink and used here.
Please do a PR adding it here
https://github.com/ArduPilot/pymavlink/blob/master/generator/mavschema.xsd
and I will merge that ASAP.

After that the `units="cs"` can be added here, and the text `Seconds * 1E2` should be replaced with `time`

hamishwillee replied:
I pushed it through because it is WIP and there is usually significant lag on getting changes in. Thanks so much for sorting this.

amilcarlucas commented on message_definitions/v1.0/common.xml:
```diff
@@ -4821,6 +4866,29 @@
+      <field type="int32_t" name="next_alt" units="mm">Next waypoint, altitude (WGS84)</field>
+      <field type="uint16_t" name="update_rate">Seconds * 1E2 until next update. Set to 0 if unknown or in data driven mode.</field>
+      <field type="uint8_t" name="flight_state" enum="UTM_FLIGHT_STATE">Flight state</field>
+      <field type="uint8_t" name="flags" enum="UTM_DATA_AVAIL_FLAGS">Bitwise OR combination of the data available flags.</field>
```
Please add display="bitmask" on this one

eyeam3 replied:
Did you see the comment above regarding the same tag? Is it then still required here?

amilcarlucas replied:
yes, it is

hamishwillee commented:
@eyeam3 As agreed in Dev meeting I'm going to merge this now. Thanks for fixing the additional feedback. Remember that this is wip so you could well have to modify it again before final acceptance :-)

eyeam3 commented:
Yes, it is WIP and I expect changes in the future both from the community and from changes in requirements.
Thanks for all the feedback!

eyeam3 commented:
@amilcarlucas I will follow up in a few minutes and make another PR.
